src/main.rs

```diff
@@ -23,14 +26,25 @@ fn main() -> Result<()> {
 
     // parse the input
     let ast = sysy::CompUnitParser::new().parse(&input).unwrap();
-    let program = ir_gen::generate_ir(&ast).unwrap();
+    let program = ir_gen::generate_on(&ast).unwrap();
 
-    // generate the output
-    let mut gen = KoopaGenerator::new(Vec::new());
-    gen.generate_on(&program).unwrap();
-    let text_form_ir = std::str::from_utf8(&gen.writer()).unwrap().to_string();
+    // generate the output conditionally
+    let text_from_ir = std::str::from_utf8(
+        // convert UTF-8 bytes to plain text
+        &(match mode.as_str() {
+            "-koopa" => {
+                let mut gen = KoopaGenerator::new(Vec::new());
+                gen.generate_on(&program).unwrap();
+                gen.writer()
+            }
+            "-riscv" => asm_gen::generate_on(&program).unwrap().into(),
+            _ => panic!("Invalid mode"),
+        }),
+    )
+    .unwrap()
+    .to_string();
```

Comment:
```suggestion
    let text_from_ir = std::str::from_utf8(
        &(match mode.as_str() {
            "-koopa" => { // generate Koopa IR using koopa::back
                let mut gen = KoopaGenerator::new(Vec::new());
                gen.generate_on(&program).unwrap();
                gen.writer()
            } // generate riscv assembly using crate::gen::asm
            "-riscv" => asm_gen::generate_on(&program).unwrap().into(),
            _ => panic!("Invalid mode"),
        }),
    ) // convert UTF-8 bytes to plain text
    .unwrap()
    .to_string();
```

Move comment to correct position.

Add more comments.

Reply:
Solved.